snuba/environment.py

```diff
@@ -31,14 +29,26 @@ def setup_sentry() -> None:
         release=os.getenv("SNUBA_RELEASE"),
     )
 
+def get_clickhouse_rw(dataset: Optional[str] = None) -> ClickhousePool:
+    if not dataset:
+        return ClickhousePool(settings.CLICKHOUSE_HOST, settings.CLICKHOUSE_PORT)
 
-clickhouse_rw = ClickhousePool(settings.CLICKHOUSE_HOST, settings.CLICKHOUSE_PORT)
-clickhouse_ro = ClickhousePool(
-    settings.CLICKHOUSE_HOST,
-    settings.CLICKHOUSE_PORT,
-    client_settings={"readonly": True},
-)
+    host = settings.CLICKHOUSE_HOST_BY_DATASET.get(dataset, settings.CLICKHOUSE_HOST)
+    port = settings.CLICKHOUSE_PORT_BY_DATASET.get(dataset, settings.CLICKHOUSE_PORT)
+    return ClickhousePool(host, port)
 
-metrics = create_metrics("snuba")
 
-reader: Reader[ClickhouseQuery] = NativeDriverReader(clickhouse_ro)
```

Comment:
Making the reader reusable was intentional. This change reverts that.
There is some context here.
https://github.com/getsentry/snuba/pull/780
I think you can decide whether to create all of them upfront or to lazily initialize the Reader and writer object but they should be reusable and not instantiated every time.

Reply:
I'd like to make this reader a property of dataset, but need to figure out how to do this and avoid the circular dependency NativeDriverReader <- Dataset <- ClickhouseQuery <- NativeDriverReader

Reply:
I haven't looked closely enough at this to be able to give any specific feedback/direction on how to get there, but if it's useful here, my thinking was that the reader and writer would eventually be associated with a storage instead of a dataset. I don't know how this plays into the broader context of joins, though.

Reply:
@lynnagara you are right, I forgot about that. ClickhouseQuery still depends on dataset till the AST is done because, guess what? `column_expr` 
https://github.com/getsentry/snuba/blob/master/snuba/clickhouse/query.py#L51
It is not there to stay but I don't think you can remove that dependency just yet. 
Sorry about the confusion.

Reply:
No worries, I "solved" this by evicting the DictQuery to a separate file.

Reply:
> I haven't looked closely enough at this to be able to give any specific feedback/direction on how to get there, but if it's useful here, my thinking was that the reader and writer would eventually be associated with a storage instead of a dataset. I don't know how this plays into the broader context of joins, though.

After more thought, I think I was wrong about this — the reader has to be able to reference multiple storages to support joins, so associating the reader with the storage is an oversimplification.

If we introduce clusters to the data model like this (highlighted section is new), it would probably make sense to associate the reader with the cluster instead to allow joins:

<img width="379" alt="Screen Shot 2020-03-25 at 2 52 42 PM" src="https://user-images.githubusercontent.com/65315/77589081-4c300d00-6ea8-11ea-8809-f4ca09ce2de1.png">

Reply:
This makes sense (the cluster idea). It is critical to figure out sooner than later how that fits in the query execution strategy. The code that loads the reader today does not know about storages (this decision happened before) https://github.com/getsentry/snuba/blob/master/snuba/web/query.py#L23.

This is something that has to be solved joins or not since the last data structure that knows about the storages is the StorageQueryPlanBuilder so there is a gap as soon as we have multiple clusters. The StorageQueryPlanBuilder though picks the execution strategy so it is meant to influence a lot of parameters of the query executions. I would start from there to see how to let the query execution code pick the right cluster.